c0x12c/spartan-module-feature-flag-kotlin
# Refactor: Convert Feature Flags to Use FeatureFlag Objects and JSON Metadata

## Summary

This PR refactors the feature flag implementation to enhance type safety, flexibility, and correct behavior:
- Replace Map with FeatureFlag objects for stronger typing
- Introduce MetadataContent sealed class for structured and type-safe metadata
- Update FeatureFlagService, Repository, and Cache to work with new FeatureFlag and MetadataContent structure
- Improve the logic in isFeatureFlagEnabled function for correct behavior
- Revise all tests to use FeatureFlag objects and validate new metadata handling

### Why

- Currently, we are using Map to represent feature flags, which reduces type safety and can lead to errors during development.
- Metadata is being stored as Map<String, String>, limiting the ability to store more complex and structured data.
- The isFeatureFlagEnabled function had a logical issue that needed to be addressed for correct behavior.
- We need a more robust and type-safe way to handle different types of feature flag targeting and configurations.

### What

- Convert from using Map to FeatureFlag objects throughout the module.
- Introduce MetadataContent sealed class with subclasses for different targeting strategies (e.g., UserTargeting, GroupTargeting, TimeBasedActivation, etc.).
- Update FeatureFlagService, FeatureFlagRepository, and related classes to work with FeatureFlag objects and MetadataContent.
- Improve the logic in isFeatureFlagEnabled to correctly handle cases with no metadata and different MetadataContent types.
- Update all unit tests and integration tests to reflect these changes and new MetadataContent structure.

### Solution


- Create a FeatureFlag data class to represent feature flags, providing better type safety and a clearer structure.
- Implement MetadataContent as a sealed class with subclasses for each targeting strategy:
> - UserTargeting
> - GroupTargeting
> - TimeBasedActivation
> - GradualRollout
> - ABTestingConfig
> - VersionTargeting
> - GeographicTargeting
> - DeviceTargeting
> - CustomRules

- 
- Update FeatureFlagService to handle different MetadataContent types:
> - Implement specific targeting logic for each MetadataContent subclass
> - Refactor isFeatureFlagEnabled to use pattern matching on MetadataContent types

- Refactor isFeatureFlagEnabled to:
> - Return false immediately if the flag is not enabled
> - Return true if the flag is enabled but has no metadata
> - Evaluate based on specific MetadataContent type if metadata is present

- Adjust FeatureFlagRepository and RedisCache to work with FeatureFlag objects and serialize/deserialize MetadataContent.
- Update the entire test suite to use FeatureFlag objects with MetadataContent and test new functionality related to each targeting strategy.

## Types of Changes



## Test Plan


- Run the entire updated test suite to ensure all functions work as expected with the new FeatureFlag objects and MetadataContent structure.
- Add specific tests for each MetadataContent subclass to verify correct targeting behavior.
- Add specific tests for isFeatureFlagEnabled to cover:
> Flags that are not enabled
> Enabled flags without metadata
> Enabled flags with various types of MetadataContent

## Checklist:
- [x] I have tested that the feature or bug fix works as expected
- [x] I have included helpful comments, particularly in hard-to-understand areas
- [x] I have added tests that prove my changes are functioning

## Related Issues

N/A
